momento-functions/examples/turbopuffer-recommended-articles.rs

```diff
@@ -0,0 +1,453 @@
+    #[serde(rename = "metadata$title")]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    metadata_title: Option<String>,
+    #[serde(rename = "metadata$link")]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    metadata_link: Option<String>,
+}
+
+// Default to 5 minute caching time for queries in Momento
+const DEFAULT_TTL_SECONDS: u64 = 300;
+// Filter out articles that aren't similar enough
+const MAXIMUM_COSINE_DISTANCE: f32 = 0.6;
+
+momento_functions::post!(get_recommended_articles);
+fn get_recommended_articles(Json(request): Json<Request>) -> WebResult<WebResponse> {
+    let headers = headers();
+    setup_logging(&headers)?;
+
+    // These are passed in as environment variables when creating the function
+    let turbopuffer_api_key = format!(
+        "Bearer {}",
+        std::env::var("TURBOPUFFER_API_KEY").unwrap_or_default()
+    );
+    let turbopuffer_endpoint = std::env::var("TURBOPUFFER_ENDPOINT").unwrap_or_default();
+    let ttl_seconds = std::env::var("TTL_SECONDS")
+        .unwrap_or(DEFAULT_TTL_SECONDS.to_string())
+        .parse::<u64>()
+        .unwrap_or(DEFAULT_TTL_SECONDS);
+    // Default to top 5 results if not provided
+    let ttl = Duration::from_secs(ttl_seconds);
+
+    let Request { article_ids, topk } = request;
+
+    // Default to top 10 articles if not provided
+    let topk = topk.unwrap_or(10);
+
+    // Get the embeddings from our provided articles
+    let embeddings = get_cached_article_embeddings(
+        article_ids.clone(),
+        &turbopuffer_endpoint,
+        &turbopuffer_api_key,
+        &ttl,
+    )?;
+
+    // Calculate the mean vector from our Vector of Vectors
+    let mean_vector = match mean_vector(&embeddings) {
+        Some(result) => result,
+        None => return Ok(WebResponse::new().with_status(500).with_body(json!({
+            "message": "Failed to calculate mean vector of embeddings, this is a bug!".to_string(),
+        }))?),
+    };
+
+    // We have our mean vector, now query Turbopuffer to find similar articles close to our mean vector,
+    // while also filtering out the articles we've already seen
+    let recommended_articles = get_similar_articles_from_turbopuffer(
+        mean_vector,
+        article_ids,
+        topk,
+        &turbopuffer_endpoint,
+        &turbopuffer_api_key,
+    )?;
+
+    let response_body = serde_json::to_vec(&recommended_articles)?;
+    Ok(WebResponse::new()
+        .with_status(200)
+        .with_headers(vec![(
+            "Content-Type".to_string(),
+            "application/json".to_string(),
+        )])
+        .with_body(response_body)?)
+}
+
+// ------------------------------------------------------
+// | Utility functions for convenience
+// ------------------------------------------------------
+
+/// Gets existing article vector embeddings from our momento cache first,
+/// then queries (and caches) any misses from Turbopuffer itself
+fn get_cached_article_embeddings(
```

Comment:
calling this `get_article_embeddings` will help for clarity. Then it should call two other functions: `get_article_embeddings_from_cache` (which returns the found embeddings and misses) and `get_article_embeddings_from_turbopuffer`

Reply:
https://github.com/momentohq/functions/pull/6/commits/7550b06c9ee6a4338f3069f1cdf6886658f4d39e